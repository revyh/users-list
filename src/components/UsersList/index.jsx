import React from 'react';
import muiThemeable from 'material-ui/styles/muiThemeable';
import Toggle from 'material-ui/Toggle';
import {List, ListItem} from 'material-ui/List';
import Subheader from 'material-ui/Subheader';
import Divider from 'material-ui/Divider';
import SortIcon from 'material-ui/svg-icons/av/sort-by-alpha';
import ArrowUpwardIcon from 'material-ui/svg-icons/navigation/arrow-upward';
import ArrowDownwardIcon from 'material-ui/svg-icons/navigation/arrow-downward';

const OrderIconMap = {
  none: SortIcon,
  ascending: ArrowUpwardIcon,
  descending: ArrowDownwardIcon,
};

const nextOrderMap = {
  none: 'ascending',
  ascending: 'descending',
  descending: 'none',
};

class UsersList extends React.Component {
  static defaultProps = {
    isGrouped: false,
    order: 'none',
  };

  handleGroupingChange = () => {
    const {isGrouped, onGroupingChange} = this.props;
    onGroupingChange && onGroupingChange(!isGrouped);
  };

  handleOrderChange = () => {
    const {order, onOrderChange} = this.props;
    onOrderChange && onOrderChange(nextOrderMap[order] || 'none');
  };

  renderUsers() {
    const {users, isGrouped} = this.props;

    return isGrouped
      ? Object.keys(users).reduce(
        (result, group) => {
          result.push(
            <Subheader key={group}>{group}</Subheader>,
            ...this.renderGroup(users[group]),
          );
          return result;
        },
        [],
      )
      : this.renderGroup(users);
  }

  renderGroup(group) {
    return group.map(
      ({id, firstName, lastName, department}) => (
        <ListItem
          key={id}
          primaryText={`${lastName} ${firstName}`}
          secondaryText={
            this.props.isGrouped
              ? null
              : department
          }
        />
      ),
    );
  }

  render() {
    const {order, isGrouped, muiTheme} = this.props;
    const OrderIcon = OrderIconMap[order] || SortIcon;

    return (
      <List style={{border: `1px solid ${muiTheme.palette.borderColor}`}}>
        <Subheader>Settings</Subheader>
        <ListItem
          key="grouping"
          primaryText="Group by department"
          rightToggle={<Toggle toggled={isGrouped} />}
          onTouchTap={this.handleGroupingChange}
        />
        <ListItem
          key="order"
          primaryText="Sort by name"
          rightIcon={<OrderIcon />}
          onTouchTap={this.handleOrderChange}
        />
        <Divider />
        {this.renderUsers()}
      </List>
    );
  }
}

export default muiThemeable()(UsersList);
