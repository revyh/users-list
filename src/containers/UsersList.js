import {connect} from 'react-redux';
import {createStructuredSelector} from 'reselect';
import {actions, selectors} from 'features/users';
import {users as usersSelector} from 'features/selectors';
import Presentational from 'components/UsersList';

const getProps = createStructuredSelector({
  users: selectors.getUsers,
  order: selectors.getOrder,
  isGrouped: selectors.getGrouping,
});

function mapStateToProps(state) {
  return getProps(usersSelector(state));
}

const mapDispatchToProps = {
  onOrderChange: actions.setUsersOrder,
  onGroupingChange: actions.setUsersGrouping,
};

export default connect(mapStateToProps, mapDispatchToProps)(Presentational);
