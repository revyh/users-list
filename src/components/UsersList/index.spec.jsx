import React from 'react';
import {shallow} from 'enzyme';
import getMuiTheme from 'material-ui/styles/getMuiTheme';
import UsersList from '.';

function shallowWithContext(node) {
  return shallow(node, {context: {muiTheme: getMuiTheme()}}).shallow();
}

function getUsers() {
  return [
    {
      id: 'someId',
      firstName: 'SomeFirstName',
      lastName: 'SomeLastName',
      department: 'Some department',
    },
    {
      id: 'anotherId',
      firstName: 'AnotherFirstName',
      lastName: 'AnotherLastName',
      department: 'Another department',
    },
  ];
}

it('should render', () => {
  expect(shallowWithContext(<UsersList users={getUsers()} />))
    .toMatchSnapshot();
});

it('should render ordered', () => {
  expect(shallowWithContext(<UsersList order="ascending" users={getUsers()} />))
    .toMatchSnapshot();

  expect(shallowWithContext(<UsersList order="descending" users={getUsers()} />))
    .toMatchSnapshot();
});

it('should render grouped', () => {
  const users = {
    'Some department': [
      {
        id: 'someId',
        firstName: 'SomeFirstName',
        lastName: 'SomeLastName',
        department: 'Some department',
      },
    ],
    'Another department': [
      {
        id: 'anotherId',
        firstName: 'AnotherFirstName',
        lastName: 'AnotherLastName',
        department: 'Another department',
      },
    ],
  };

  expect(shallowWithContext(<UsersList isGrouped users={users} />))
    .toMatchSnapshot();
});

it('should trigger \'onGroupingChange\' event handler', () => {
  const mockOnGroupingChange = jest.fn();
  const usersList = shallowWithContext(
    <UsersList users={getUsers()} onGroupingChange={mockOnGroupingChange} />,
  );

  usersList.findWhere(
    node => node.name() === 'ListItem' && node.key() === 'grouping',
  ).simulate('touchTap');

  expect(mockOnGroupingChange).toHaveBeenCalledWith(true);
});

it('should trigger \'onOrderChange\' event handler', () => {
  const mockOnOrderChange = jest.fn();
  const usersList = shallowWithContext(
    <UsersList users={getUsers()} onOrderChange={mockOnOrderChange} />,
  );

  const orderItem = usersList.findWhere(
    node => node.name() === 'ListItem' && node.key() === 'order',
  );

  orderItem.simulate('touchTap');
  expect(mockOnOrderChange).toHaveBeenCalledWith('ascending');
  mockOnOrderChange.mockClear();
  usersList.setProps({order: 'ascending'});

  orderItem.simulate('touchTap');
  expect(mockOnOrderChange).toHaveBeenCalledWith('descending');
  mockOnOrderChange.mockClear();
  usersList.setProps({order: 'descending'});

  orderItem.simulate('touchTap');
  expect(mockOnOrderChange).toHaveBeenCalledWith('none');
});
