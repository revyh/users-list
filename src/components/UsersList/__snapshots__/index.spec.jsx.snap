// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<List
  style={
    Object {
      "border": "1px solid #e0e0e0",
    }
  }
>
  <Subheader
    inset={false}
  >
    Settings
  </Subheader>
  <ListItem
    autoGenerateNestedIndicator={true}
    containerElement="span"
    disableKeyboardFocus={false}
    disabled={false}
    initiallyOpen={false}
    insetChildren={false}
    nestedItems={Array []}
    nestedLevel={0}
    onKeyboardFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onNestedListToggle={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    onTouchTap={[Function]}
    open={null}
    primaryText="Group by department"
    primaryTogglesNestedList={false}
    rightToggle={
      <Toggle
        defaultToggled={false}
        disabled={false}
        labelPosition="left"
        toggled={false}
      />
    }
    secondaryTextLines={1}
  />
  <ListItem
    autoGenerateNestedIndicator={true}
    containerElement="span"
    disableKeyboardFocus={false}
    disabled={false}
    initiallyOpen={false}
    insetChildren={false}
    nestedItems={Array []}
    nestedLevel={0}
    onKeyboardFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onNestedListToggle={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    onTouchTap={[Function]}
    open={null}
    primaryText="Sort by name"
    primaryTogglesNestedList={false}
    rightIcon={<AvSortByAlpha />}
    secondaryTextLines={1}
  />
  <Divider
    inset={false}
  />
  <ListItem
    autoGenerateNestedIndicator={true}
    containerElement="span"
    disableKeyboardFocus={false}
    disabled={false}
    initiallyOpen={false}
    insetChildren={false}
    nestedItems={Array []}
    nestedLevel={0}
    onKeyboardFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onNestedListToggle={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    open={null}
    primaryText="SomeLastName SomeFirstName"
    primaryTogglesNestedList={false}
    secondaryText="Some department"
    secondaryTextLines={1}
  />
  <ListItem
    autoGenerateNestedIndicator={true}
    containerElement="span"
    disableKeyboardFocus={false}
    disabled={false}
    initiallyOpen={false}
    insetChildren={false}
    nestedItems={Array []}
    nestedLevel={0}
    onKeyboardFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onNestedListToggle={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    open={null}
    primaryText="AnotherLastName AnotherFirstName"
    primaryTogglesNestedList={false}
    secondaryText="Another department"
    secondaryTextLines={1}
  />
</List>
`;

exports[`should render grouped 1`] = `
<List
  style={
    Object {
      "border": "1px solid #e0e0e0",
    }
  }
>
  <Subheader
    inset={false}
  >
    Settings
  </Subheader>
  <ListItem
    autoGenerateNestedIndicator={true}
    containerElement="span"
    disableKeyboardFocus={false}
    disabled={false}
    initiallyOpen={false}
    insetChildren={false}
    nestedItems={Array []}
    nestedLevel={0}
    onKeyboardFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onNestedListToggle={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    onTouchTap={[Function]}
    open={null}
    primaryText="Group by department"
    primaryTogglesNestedList={false}
    rightToggle={
      <Toggle
        defaultToggled={false}
        disabled={false}
        labelPosition="left"
        toggled={true}
      />
    }
    secondaryTextLines={1}
  />
  <ListItem
    autoGenerateNestedIndicator={true}
    containerElement="span"
    disableKeyboardFocus={false}
    disabled={false}
    initiallyOpen={false}
    insetChildren={false}
    nestedItems={Array []}
    nestedLevel={0}
    onKeyboardFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onNestedListToggle={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    onTouchTap={[Function]}
    open={null}
    primaryText="Sort by name"
    primaryTogglesNestedList={false}
    rightIcon={<AvSortByAlpha />}
    secondaryTextLines={1}
  />
  <Divider
    inset={false}
  />
  <Subheader
    inset={false}
  >
    Some department
  </Subheader>
  <ListItem
    autoGenerateNestedIndicator={true}
    containerElement="span"
    disableKeyboardFocus={false}
    disabled={false}
    initiallyOpen={false}
    insetChildren={false}
    nestedItems={Array []}
    nestedLevel={0}
    onKeyboardFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onNestedListToggle={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    open={null}
    primaryText="SomeLastName SomeFirstName"
    primaryTogglesNestedList={false}
    secondaryText={null}
    secondaryTextLines={1}
  />
  <Subheader
    inset={false}
  >
    Another department
  </Subheader>
  <ListItem
    autoGenerateNestedIndicator={true}
    containerElement="span"
    disableKeyboardFocus={false}
    disabled={false}
    initiallyOpen={false}
    insetChildren={false}
    nestedItems={Array []}
    nestedLevel={0}
    onKeyboardFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onNestedListToggle={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    open={null}
    primaryText="AnotherLastName AnotherFirstName"
    primaryTogglesNestedList={false}
    secondaryText={null}
    secondaryTextLines={1}
  />
</List>
`;

exports[`should render ordered 1`] = `
<List
  style={
    Object {
      "border": "1px solid #e0e0e0",
    }
  }
>
  <Subheader
    inset={false}
  >
    Settings
  </Subheader>
  <ListItem
    autoGenerateNestedIndicator={true}
    containerElement="span"
    disableKeyboardFocus={false}
    disabled={false}
    initiallyOpen={false}
    insetChildren={false}
    nestedItems={Array []}
    nestedLevel={0}
    onKeyboardFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onNestedListToggle={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    onTouchTap={[Function]}
    open={null}
    primaryText="Group by department"
    primaryTogglesNestedList={false}
    rightToggle={
      <Toggle
        defaultToggled={false}
        disabled={false}
        labelPosition="left"
        toggled={false}
      />
    }
    secondaryTextLines={1}
  />
  <ListItem
    autoGenerateNestedIndicator={true}
    containerElement="span"
    disableKeyboardFocus={false}
    disabled={false}
    initiallyOpen={false}
    insetChildren={false}
    nestedItems={Array []}
    nestedLevel={0}
    onKeyboardFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onNestedListToggle={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    onTouchTap={[Function]}
    open={null}
    primaryText="Sort by name"
    primaryTogglesNestedList={false}
    rightIcon={<NavigationArrowUpward />}
    secondaryTextLines={1}
  />
  <Divider
    inset={false}
  />
  <ListItem
    autoGenerateNestedIndicator={true}
    containerElement="span"
    disableKeyboardFocus={false}
    disabled={false}
    initiallyOpen={false}
    insetChildren={false}
    nestedItems={Array []}
    nestedLevel={0}
    onKeyboardFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onNestedListToggle={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    open={null}
    primaryText="SomeLastName SomeFirstName"
    primaryTogglesNestedList={false}
    secondaryText="Some department"
    secondaryTextLines={1}
  />
  <ListItem
    autoGenerateNestedIndicator={true}
    containerElement="span"
    disableKeyboardFocus={false}
    disabled={false}
    initiallyOpen={false}
    insetChildren={false}
    nestedItems={Array []}
    nestedLevel={0}
    onKeyboardFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onNestedListToggle={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    open={null}
    primaryText="AnotherLastName AnotherFirstName"
    primaryTogglesNestedList={false}
    secondaryText="Another department"
    secondaryTextLines={1}
  />
</List>
`;

exports[`should render ordered 2`] = `
<List
  style={
    Object {
      "border": "1px solid #e0e0e0",
    }
  }
>
  <Subheader
    inset={false}
  >
    Settings
  </Subheader>
  <ListItem
    autoGenerateNestedIndicator={true}
    containerElement="span"
    disableKeyboardFocus={false}
    disabled={false}
    initiallyOpen={false}
    insetChildren={false}
    nestedItems={Array []}
    nestedLevel={0}
    onKeyboardFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onNestedListToggle={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    onTouchTap={[Function]}
    open={null}
    primaryText="Group by department"
    primaryTogglesNestedList={false}
    rightToggle={
      <Toggle
        defaultToggled={false}
        disabled={false}
        labelPosition="left"
        toggled={false}
      />
    }
    secondaryTextLines={1}
  />
  <ListItem
    autoGenerateNestedIndicator={true}
    containerElement="span"
    disableKeyboardFocus={false}
    disabled={false}
    initiallyOpen={false}
    insetChildren={false}
    nestedItems={Array []}
    nestedLevel={0}
    onKeyboardFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onNestedListToggle={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    onTouchTap={[Function]}
    open={null}
    primaryText="Sort by name"
    primaryTogglesNestedList={false}
    rightIcon={<NavigationArrowDownward />}
    secondaryTextLines={1}
  />
  <Divider
    inset={false}
  />
  <ListItem
    autoGenerateNestedIndicator={true}
    containerElement="span"
    disableKeyboardFocus={false}
    disabled={false}
    initiallyOpen={false}
    insetChildren={false}
    nestedItems={Array []}
    nestedLevel={0}
    onKeyboardFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onNestedListToggle={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    open={null}
    primaryText="SomeLastName SomeFirstName"
    primaryTogglesNestedList={false}
    secondaryText="Some department"
    secondaryTextLines={1}
  />
  <ListItem
    autoGenerateNestedIndicator={true}
    containerElement="span"
    disableKeyboardFocus={false}
    disabled={false}
    initiallyOpen={false}
    insetChildren={false}
    nestedItems={Array []}
    nestedLevel={0}
    onKeyboardFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onNestedListToggle={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    open={null}
    primaryText="AnotherLastName AnotherFirstName"
    primaryTogglesNestedList={false}
    secondaryText="Another department"
    secondaryTextLines={1}
  />
</List>
`;
